### DISCUSSION TITLE
Error Building WASM

### DISCUSSION THREAD

**Discussion by Loosetooth** [elalish/manifold#590] [Q&A] [answered]
I just freshly cloned (with subrepos) and built manifold, which worked fine.

I then planned to move on to the WASM build, because I'm interested in making an NPM package for the newest version.
Following the commands in the readme file, running the `emcmake cmake -DCMAKE_BUILD_TYPE=Release .. && emmake make` results in the following error on my machine:

```
CMake Error at buildWASM/_deps/thrust-src/cmake/ThrustAddSubdir.cmake:1 (find_package):
  Could not find a package configuration file provided by "Thrust" with any
  of the following names:

    ThrustConfig.cmake
    thrust-config.cmake

  Add the installation prefix of "Thrust" to CMAKE_PREFIX_PATH or set
  "Thrust_DIR" to a directory containing one of the above files.  If "Thrust"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  buildWASM/_deps/thrust-src/CMakeLists.txt:59 (include)
```

I checked the `buildWASM` folder, and it has the thrust-src repo cloned with everything in there:

![image](https://github.com/elalish/manifold/assets/66420755/96412118-d4bb-4f56-86f2-4be89b42102f)

I'm not a CMAKE expert, so I don't really know what to do about this error. Any ideas?

**Comment by pca006132** [comment]
hmmm this is weird, we had this in the CI: https://github.com/elalish/manifold/blob/master/.github/workflows/manifold.yml#L91

Can you use a docker container to reproduce it and give me the list of commands that you used?

**Reply by Loosetooth** [reply]
## local machine
Regarding the steps and list of commands that I used:
1. I made a fresh clone of the manifold repo, including subrepos, so everything should have been up-to-date.
2. I followed the [WASM steps as shown in the manifold readme](https://github.com/elalish/manifold#wasm) for linux:
```
git clone https://github.com/emscripten-core/emsdk.git
cd emsdk
./emsdk install latest
./emsdk activate latest
source ./emsdk/emsdk_env.sh
```
And then:
```
cd manifold
mkdir buildWASM
cd buildWASM
emcmake cmake -DCMAKE_BUILD_TYPE=Release .. && emmake make
```
3. tried different nodejs versions, same error.
4. made sure emscripten was latest version, same error.

## Docker
Now let's try in docker (sorry for the long description, this is how I did it):

1. Start ubuntu:22.04 container and keep running: `docker run -d -t --name manifold-wasm-build ubuntu:22.04`
2. find container id using `docker ps`
3. Open interactive terminal: `docker exec -it 1e986469139d /bin/bash`
4. Prepare:
```
apt-get update
apt-get install -y git build-essential cmake nodejs
apt-get install -y software-properties-common gcc && \
    add-apt-repository -y ppa:deadsnakes/ppa
apt-get install -y python3.8
git clone --recurse-submodules https://github.com/elalish/manifold.git
git clone https://github.com/emscripten-core/emsdk.git
cd emsdk
./emsdk install latest
./emsdk activate latest
cd ..
source ./emsdk/emsdk_env.sh
```
Then build:
```
cd manifold
mkdir buildWASM
cd buildWASM
emcmake cmake -DCMAKE_BUILD_TYPE=Release .. && emmake make
```

Observe the error:
```
emcmake cmake -DCMAKE_BUILD_TYPE=Release .. && emmake make
configure: cmake -DCMAKE_BUILD_TYPE=Release .. -DCMAKE_TOOLCHAIN_FILE=/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR=/emsdk/node/16.20.0_64bit/bin/node
Building for Emscripten
Parallel Backend: NONE
CMake Error at buildWASM/_deps/thrust-src/cmake/ThrustAddSubdir.cmake:1 (find_package):
  Could not find a package configuration file provided by "Thrust" with any
  of the following names:

    ThrustConfig.cmake
    thrust-config.cmake

  Add the installation prefix of "Thrust" to CMAKE_PREFIX_PATH or set
  "Thrust_DIR" to a directory containing one of the above files.  If "Thrust"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  buildWASM/_deps/thrust-src/CMakeLists.txt:59 (include)


-- Configuring incomplete, errors occurred!
See also "/manifold/buildWASM/CMakeFiles/CMakeOutput.log".
emcmake: error: 'cmake -DCMAKE_BUILD_TYPE=Release .. -DCMAKE_TOOLCHAIN_FILE=/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR=/emsdk/node/16.20.0_64bit/bin/node' failed (returned 1)
```

**Comment by t-paul** [comment]
I'm seeing the same error in OpenSCAD builds: https://app.circleci.com/pipelines/github/openscad/openscad/4144/workflows/7e09c516-24fd-463a-9d2b-94e1571b565c/jobs/16738

We have a `-DTHRUST_HOST_SYSTEM=THRUST_HOST_SYSTEM_TBB`, could that interfere with the latest changes maybe?

**Comment by pca006132** [comment, marked as answer]
See #591. Thanks for letting us know this issue!
